### ISSUE TITLE
Tensorboard does not support Python 3

### ISSUE THREAD

**Issue by dankolbman** [tensorflow/tensorflow#470]
[tensorboard.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tensorboard/tensorboard.py) and [tensorboard_handler.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tensorboard/tensorboard_handler.py) are still using `BaseHTTPServer` which was move to `http.server` in Python 3. There are some other changes with `StringIO` and `urlparse` that were quickly resolved with `2to3`.

Strings also need to be encoded for the `wfile.write`s in [tensorboard_handler.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tensorboard/tensorboard_handler.py).

Everything seemed to work after I made these changes.

**Comment by vrv** [comment]
Thanks -- We'll try to update tensorboard to be python 3 compatible -- will probably be in the next binary release.

**Comment by Kasmetski** [comment]
nice

**Comment by kuangray** [comment]
To dankolbman,

I ran into the same issue. Would you like to share your modified tensorboard.py and tensorboard_handler.py for working with python3? Thanks in advance!

**Comment by dankolbman** [comment]
Yeah sure, here are the diffs. `2to3` will take care of everything except for the encoding issue on line 171 of `tensorboard_handler.py`. Just replace it with `self.wfile.write(output.encode('utf-8'))`

`tensorboard.py`:
```diff
-import http.server
+import BaseHTTPServer
 import functools
 import os
 import socket
-import socketserver
+import SocketServer
 
@@ -107,8 +107,8 @@ def ParseEventFilesFlag(flag_value):
-class ThreadedHTTPServer(socketserver.ThreadingMixIn,
-                         http.server.HTTPServer):
+class ThreadedHTTPServer(SocketServer.ThreadingMixIn,
+                         BaseHTTPServer.HTTPServer):
```

`tensorboard_handler.py`:
```diff
-import http.server
+import BaseHTTPServer
-import io
-import urllib.parse
+import StringIO
+import urlparse
 
@@ -76,7 +76,7 @@ class _OutputFormat(object):
-class TensorboardHandler(http.server.BaseHTTPRequestHandler):
+class TensorboardHandler(BaseHTTPServer.BaseHTTPRequestHandler):
   """Handler class for use with BaseHTTPServer.HTTPServer.

@@ -85,7 +85,7 @@ class TensorboardHandler(http.server.BaseHTTPRequestHandler):
   def __init__(self, multiplexer, *args):
     self._multiplexer = multiplexer
-    http.server.BaseHTTPRequestHandler.__init__(self, *args)
+    BaseHTTPServer.BaseHTTPRequestHandler.__init__(self, *args)

@@ -140,7 +140,7 @@ class TensorboardHandler(http.server.BaseHTTPRequestHandler):
-    out = io.StringIO()
+    out = StringIO.StringIO()
     f = gzip.GzipFile(fileobj=out, mode='w')

@@ -168,7 +168,7 @@ class TensorboardHandler(http.server.BaseHTTPRequestHandler):
     self.send_header('Content-Type', 'application/json')
     self.send_header('Content-Length', len(output))
     self.end_headers()
-    self.wfile.write(output.encode('utf-8'))
+    self.wfile.write(output)

@@ -195,7 +195,7 @@ class TensorboardHandler(http.server.BaseHTTPRequestHandler):
     values = self._multiplexer.Scalars(run, tag)
 
     if query_params.get('format') == _OutputFormat.CSV:
-      string_io = io.StringIO()
+      string_io = StringIO.StringIO()
       writer = csv.writer(string_io)

@@ -234,7 +234,7 @@ class TensorboardHandler(http.server.BaseHTTPRequestHandler):
     run = query_params.get('run')
     compressed_histograms = self._multiplexer.CompressedHistograms(run, tag)
     if query_params.get('format') == _OutputFormat.CSV:
-      string_io = io.StringIO()
+      string_io = StringIO.StringIO()
       writer = csv.writer(string_io)

@@ -242,7 +242,7 @@ class TensorboardHandler(http.server.BaseHTTPRequestHandler):
       headers = ['Wall time', 'Step']
       if compressed_histograms:
         bucket_count = len(compressed_histograms[0].compressed_histogram_values)
-        for i in range(bucket_count):
+        for i in xrange(bucket_count):
           headers += ['Edge %d basis points' % i, 'Edge %d value' % i]
       writer.writerow(headers)
 
@@ -366,7 +366,7 @@ class TensorboardHandler(http.server.BaseHTTPRequestHandler):
   def do_GET(self):  # pylint: disable=invalid-name
     """Handler for all get requests."""
-    parsed_url = urllib.parse.urlparse(self.path)
+    parsed_url = urlparse.urlparse(self.path)

@@ -386,7 +386,7 @@ class TensorboardHandler(http.server.BaseHTTPRequestHandler):
     if clean_path in handlers:
-      query_params = urllib.parse.parse_qs(parsed_url.query)
+      query_params = urlparse.parse_qs(parsed_url.query)
       # parse_qs returns a list of values for each key; we're only interested in
```

**Comment by kuangray** [comment]
Thanks! The updated script works. My 2to3 did not do a good job as this.

**Comment by kibtes** [comment]
Finally Tensorboard worked with python3 but now the issue is that the graph failed to be visual. I tried to visualize it on chrome.Can somebody please help?
The error is 
Graph Visualization failed:Error:Failure Parsing graph definition

**Comment by windeveloper** [comment]
I run into the same error with python3 

Graph Visualization failed:Error:Failure Parsing graph definition

**Comment by nikitakit** [comment]
I ran into graph visualization errors as well. The fix is to change `f.write(content)` to `f.write(content.encode('utf-8'))` in `_send_gzip_response`

**Comment by DSLituiev** [comment]
@dankolbman , will you consider making a pull request with your diffs?

**Comment by dankolbman** [comment]
Unfortunately, the changes here are just a hack and break compatability with python2. Some more work needs to be done to work with both python2 and 3.

**Comment by kibtes** [comment]
To nikitakit,

i included your fix to the fix suggested by dankolbman but i still have the same error.

Type Error: string argument expected,got 'bytes'
If any other ideas please help!

**Comment by martinwicke** [comment]
You probably have to replace StringIO with BytesIO somewhere. I've seen
this kind of error before here
<http://stackoverflow.com/questions/34416702/tensorboard-graph-visualiyation-error-using-python-3/34459035#34459035>.
Is that the same error you're getting?

**Comment by nikitakit** [comment]
That's right, `_send_gzip_response` should use BytesIO, not StringIO. Sorry, I didn't run a full diff and missed that.

```diff
-    out = StringIO.StringIO()
+    out = io.BytesIO()
     f = gzip.GzipFile(fileobj=out, mode='w')
-    f.write(content)
+    f.write(content.encode('utf-8'))
```

**Comment by frankyjuang** [comment]
I fix all the bugs mentioned above and there's no error on the server side now. However, besides top bar and side column, there's only "Main Graph" two words on the graph page. Any ideas please?
![image](https://cloud.githubusercontent.com/assets/6175880/12097756/72574918-b358-11e5-9ee7-285a6d05deb5.png)

ps. what I'm doing is just import the image recognition model (google inception v3) and summary write the graph def to file trying to see how does the model look like. Thanks...

**Comment by nikitakit** [comment]
Are you using chrome?

I get that blank page in both firefox and safari, but chrome works. Not sure if this is because of something else missing regarding python3, or just a tensorboard bug.

**Comment by kibtes** [comment]
@martinwicke 
yes it is the same error. I just tried to fix with the suggestion posted by @nikitakit and i have error. 
The Error:
Exception happened during processing of request from ('127.0.0.1', 62162)
Traceback (most recent call last):
  File "/usr/lib/python3.4/socketserver.py", line 617, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python3.4/socketserver.py", line 344, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.4/site-packages/tensorflow/tensorboard/tensorboard_handler.py", line 86, in __init__
    http.server.BaseHTTPRequestHandler.__init__(self,*args)#BaseHTTPServer.BaseHTTPRequestHandler.__init__(self, *args)
  File "/usr/lib/python3.4/socketserver.py", line 673, in __init__
    self.handle()
  File "/usr/lib/python3.4/http/server.py", line 398, in handle
    self.handle_one_request()
  File "/usr/lib/python3.4/http/server.py", line 386, in handle_one_request
    method()
  File "/usr/lib/python3.4/site-packages/tensorflow/tensorboard/tensorboard_handler.py", line 382, in do_GET
    handlers[clean_path](query_params)
  File "/usr/lib/python3.4/site-packages/tensorflow/tensorboard/tensorboard_handler.py", line 210, in _serve_graph
    self._send_gzip_response(graph_pbtxt, 'text/plain')
  File "/usr/lib/python3.4/site-packages/tensorflow/tensorboard/tensorboard_handler.py", line 136, in _send_gzip_response
    f = gzip.GzipFile(fileobj=out, mode='w')
  File "/usr/lib/python3.4/gzip.py", line 220, in __init__
    self._write_gzip_header()
  File "/usr/lib/python3.4/gzip.py", line 267, in _write_gzip_header
    self.fileobj.write(chr(flags.encode('latin-1')))
AttributeError: 'int' object has no attribute 'encode'

what do I do?

**Comment by nikitakit** [comment]
@kibtes 
The erroneous line in the traceback seems suspect. On my computer, it is `self.fileobj.write(chr(flags).encode('latin-1'))` (note the different placement of quotes). And this is inside the Python installation! Did you manually modify that file, by any chance?

**Comment by kibtes** [comment]
@nikitakit 

Thanks. The fix works. I guess i had manually modified it.

**Comment by frankyjuang** [comment]
@nikitakit, yes I'm using Chromium 47.0

**Comment by jfsantos** [comment]
I updated the files in the binary release (pretty much everything that was discussed in this issue) and it works. I then tried to fix the current master by doing the same, but now no plots or graphs are shown (instead, it shows the usual messages for when the logs do not have any useful info to be shown). The only thing I've seen that seems wrong in the server logs are these messages:

```
WARNING:tensorflow:IOError [Errno 2] No such file or directory: '/usr/local/lib/python2.7/dist-packages/tensorflow/tensorboard/TAG' on path /usr/local/lib/python2.7/dist-packages/tensorflow/tensorboard/TAG
WARNING:tensorflow:Unable to read TensorBoard tag
```

(note: Python is 2.7 above because I was testing if the fixes were portable)

Any ideas on what could be going on?

**Comment by jfsantos** [comment]
I have just submitted a PR fixing this for the master branch: #749. The same fixes can be applied to 0.6.0, and I could submit a similar PR to fix those if there's interest (I don't know if the 0.6.0 binaries receive updates or if we have to wait for the next release).

**Comment by danmane** [comment]
Thanks for the PR! Since it's merged now, I'm going to close this issue.
